### ISSUE TITLE
Decode Error when extracting protobuf message (macOS)

### ISSUE THREAD

**Issue by lwoloszy** [stan-dev/pystan-next#163]
Just playing around with the package, I find that beyond a certain number of data or parameters, I seem to run into a Decode Error.

Taking the sample model...

```import stan

schools_code = """
data {
  int<lower=0> J;         // number of schools
  real y[J];              // estimated treatment effects
  real<lower=0> sigma[J]; // standard error of effect estimates
}
parameters {
  real mu;                // population treatment effect
  real<lower=0> tau;      // standard deviation in treatment effects
  vector[J] eta;          // unscaled deviation from mu by school
}
transformed parameters {
  vector[J] theta = mu + tau * eta;        // school treatment effects
}
model {
  target += normal_lpdf(eta | 0, 1);       // prior log-density
  target += normal_lpdf(y | theta, sigma); // log-likelihood
}
"""
```

This succeeds: 
```
schools_data = {"J": 8,
                "y": [28,  8, -3,  7, -1,  1, 18, 12],
                "sigma": [15, 10, 16, 11,  9, 11, 10, 18]}

posterior = stan.build(schools_code, data=schools_data)
fit = posterior.sample(num_chains=4, num_samples=1000)
```

This succeeds:
```
schools_data = {"J": 8*10,
                "y": [28,  8, -3,  7, -1,  1, 18, 12]*10,
                "sigma": [15, 10, 16, 11,  9, 11, 10, 18]*10}

posterior = stan.build(schools_code, data=schools_data)
fit = posterior.sample(num_chains=4, num_samples=1000)
```

But this fails:
```
schools_data = {"J": 8*11,
                "y": [28,  8, -3,  7, -1,  1, 18, 12]*11,
                "sigma": [15, 10, 16, 11,  9, 11, 10, 18]*11}

posterior = stan.build(schools_code, data=schools_data)
fit = posterior.sample(num_chains=4, num_samples=1000)
```

Stack trace:
```
DecodeError                               Traceback (most recent call last)
<ipython-input-3-815b881aecc2> in <module>
      1 posterior = stan.build(schools_code, data=schools_data)
----> 2 fit = posterior.sample(num_chains=1, num_samples=1000)

~/.pyenv/versions/3.8.5/envs/stan/lib/python3.8/site-packages/stan/model.py in sample(self, **kwargs)
    239
    240         try:
--> 241             return asyncio.run(go())
    242         except KeyboardInterrupt:
    243             pass

~/.pyenv/versions/3.8.5/lib/python3.8/asyncio/runners.py in run(main, debug)
     41         events.set_event_loop(loop)
     42         loop.set_debug(debug)
---> 43         return loop.run_until_complete(main)
     44     finally:
     45         try:

~/.pyenv/versions/3.8.5/lib/python3.8/asyncio/base_events.py in run_until_complete(self, future)
    614             raise RuntimeError('Event loop stopped before Future completed.')
    615
--> 616         return future.result()
    617
    618     def stop(self):

~/.pyenv/versions/3.8.5/envs/stan/lib/python3.8/site-packages/stan/model.py in go()
    188                         if resp.status != 200:
    189                             raise RuntimeError((await resp.json())["message"])
--> 190                         stan_outputs.append(tuple(extract_protobuf_messages(await resp.read())))
    191
    192                 def is_nonempty_logger_message(msg):

~/.pyenv/versions/3.8.5/envs/stan/lib/python3.8/site-packages/stan/model.py in extract_protobuf_messages(fit_bytes)
    131                 msg = callbacks_writer_pb2.WriterMessage()
    132                 next_pos, pos = varint_decoder(fit_bytes, pos)
--> 133                 msg.ParseFromString(fit_bytes[pos : pos + next_pos])
    134                 yield msg
    135                 pos += next_pos

DecodeError: Error parsing message
```

Running on MacOS 10.15.6 with 32GB, Python 3.8.5, stan 3.0.0b4, httpstan 2.3.0

**Comment by riddell-stan** [comment]
Thank you for the detailed report. I'll see if I can reproduce this on macOS.

**Comment by ahartikainen** [comment]
Should we decode the msg before it goes to ParseFromString?

    msg.ParseFromString(fit_bytes[pos : pos + next_pos].decode("utf-8"))

**Comment by riddell-stan** [comment]
This may take some time to fix/pin down. No developer working on PyStan has easy access to a macOS machine.

**Comment by lwoloszy** [comment]
FWIW, I get the same error on an AWS instance running Amazon Linux 4.14.133-88.105.amzn1.x86_64

**Comment by riddell-stan** [comment]
AWS might have a different explanation. gcc 9 is required to use the wheels. Amazon Linux 4.14 uses an earlier version, right?

pystan works fine with earlier versions of gcc, but one needs to build httpstan from source, see https://httpstan.readthedocs.io/en/latest/installation.html

**Comment by riddell-stan** [comment]
I got access to a mac. I'm able to reproduce this. I'll dig deeper this weekend.

We should certainly add this as a test case (to httpstan).

**Comment by ahartikainen** [comment]
So, it appears that part of the fit bytes is somehow malformed (some responses are not full).

This happens both in macOS and Linux (randomly).

**Comment by riddell-stan** [comment]
@ahartikainen I'm not confident there is sufficient evidence to conclude that the linux problem (if one exists) is the same as the problem on macOS. It may be. This issue is going to be about the non-randomly-occuring macOS bug.

Added a test case in httpstan to make sure we catch this in the future, https://github.com/stan-dev/httpstan/pull/490

**Comment by ahartikainen** [comment]
It might be different

https://github.com/ahartikainen/run_custom_code

Here is one run where it fails with Ubuntu https://github.com/ahartikainen/run_custom_code/runs/1334128557?check_suite_focus=true

**Comment by riddell-stan** [comment]
Worked on this a bit today. Clearly something has gone wrong with the protobuf encoding.

When I have the time, I'll try doing what the error message suggests, switching from "string" to "bytes" in our protobuf definition file, https://github.com/stan-dev/httpstan/blob/main/protos/callbacks_writer.proto

Looking at the protobuf docs, it seems like we are, in fact, doing the wrong thing. A C++ `std::string` is not a protobuf `string`.

**Comment by riddell-stan** [comment]
@lwoloszy What version of Xcode are you using?

**Comment by lwoloszy** [comment]
11.7

**Comment by riddell-stan** [comment]
Thanks.

I'll admit I do not understand what is happening. I'm really surprised that different byte sequences are being written to disk on macOS vs Linux. Protobuf is supposed to provide cross-platform serialization.

My only thought is that something is going wrong with the length-prefix encoding which is not a standard feature of protobuf (although it is in the official Java version).

My proposed "solution" to this is to move away from protobuf and use JSON instead. This switch had been planned for some time.

**Comment by riddell-stan** [comment]
Fixed in 3.0.0b5.

@lwoloszy Thanks again for the report. Sorry it took 22 days to fix. The problem turned out to be a coding error I made in Python. The reason this occurred on macos and not linux is that macos has a much smaller receive and send buffer for unix domain sockets. (`httpstan` uses unix domain sockets to send messages from C++ to Python.)
